### ISSUE TITLE
nodejs app failing with "cp: cannot stat ‘node_modules’: No such file or directory"

### ISSUE THREAD

**Issue by drnic** [cloudfoundry/nodejs-buildpack#140]
What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?

Deploying with Concourse CI 3.9.2 and its `cf-resource`.

What version of the buildpack you are using?

PWS is now running v1.6.22

The app successfully deployed to v1.6.21

If you were attempting to accomplish a task, what was it you were attempting to do?

Deploy a nodejs app https://github.com/starkandwayne/ghost-for-cloudfoundry

What was the actual behavior?

```
2018-04-16T22:07:16.98+0000 [APP/PROC/WEB/0]OUT Setting up symlinks
2018-04-16T22:07:16.98+0000 [APP/PROC/WEB/0]OUT /home/vcap/app
2018-04-16T22:07:16.98+0000 [APP/PROC/WEB/0]ERR cp: cannot stat ‘node_modules’: No such file or directory
2018-04-16T22:07:16.99+0000 [APP/PROC/WEB/0]OUT Exit status 1
```

Can you provide a sample app?

https://github.com/starkandwayne/ghost-for-cloudfoundry
Please confirm where necessary:
* [x] I have included a log output
* [x] My log includes an error message
* [x] I have included steps for reproduction

**Comment by sclevine** [comment]
Thanks for the report @drnic. We will investigate ASAP.

v1.6.21 introduced regressions that were fixed in v1.6.22. I'm guessing we broke something else.

**Comment by drnic** [comment]
TIL I can put the full buildpack URL in the manifest.yml:
```Updating app with these attributes...
  name:                ghost
  path:                /Users/drnic/Projects/starkandwayne/ghost-for-cloudfoundry
- buildpack:           nodejs_buildpack
+ buildpack:           https://github.com/cloudfoundry/nodejs-buildpack/releases/download/v1.6.21/nodejs-buildpack-v1.6.21.zip```

**Comment by jfeeny** [comment]
@drnic It looks like your app's `setup_and_run.sh` currently assumes that `node_modules` is always at the root of the application directory. This was always true for the previous buildpack version. However, in 1.6.22, it depends on whether your app is "vendored" (you're uploading `node_modules` with your app). If your app is vendored, we keep `node_modules` where it is. If your app is not vendored, `node_modules` will be found in the buildpack's dependency directory and the buildpack will set the `$NODE_PATH` environment variable accordingly.

In order for your application to be agnostic of the `node_modules` location, we recommend changing `cp -r node_modules current/` in your `setup_and_run.sh` to `cp -r ${NODE_PATH:-$HOME/node_modules} current/`.

**Comment by jfeeny** [comment]
@sclevine Currently we only set `$NODE_PATH` when the application is **not** vendored. Do you think there's a case for setting it regardless, so users can always use `$NODE_HOME`? Then the above command could become `cp -r $NODE_PATH current/`.

**Comment by sclevine** [comment]
@jfeeny I agree that we should always set `$NODE_PATH` at launch. I'll prioritize this issue to make that change.

@drnic Can you update [this line](https://github.com/starkandwayne/ghost-for-cloudfoundry/blob/master/bin/setup_and_run.sh#L146) to use `$NODE_PATH` instead of assuming that `/home/vcap/app/node_modules` exists?

It's difficult for us to avoid breaking changes in cases where the language tools allow you to execute arbitrary code during the staging process. In this case, we incorrectly assumed we were safe to make this change because:
1. It results in less modification to the app directory and
2. It used a standard environment variable (`$NODE_PATH`) to tell Node.js where `node_modules` should live.

We will document this change and consider an alternative implementation if we get more reports.

PS. You may also want to change [this line](https://github.com/starkandwayne/ghost-for-cloudfoundry/blob/master/bin/setup_and_run.sh#L149) as it makes assumptions that are not guaranteed (ex. the buildpack will always be at index 0). If the buildpack doesn't handle that for you already, it should. I'll make a note about it in the story as well.

**Comment by sclevine** [comment]
Apologies, I incorrectly assumed that you're `bin/setup_and_run.sh` script was running at build-time via a hook in `package.json`. I just realized that it's actually your start command.

We only guarantee that start commands set by the buildpack will continue to work in future buildpack versions. We make no attempt to support custom start commands. The part about arbitrary code running during the staging process is not applicable in this case.
